contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,378 @@
+     * @param options is the address of the options contract
+     * @param amount is the amount of tokens or options contract to exercise. Only relevant for fungle protocols like Opyn
+     */
+    function exerciseProfit(
+        address options,
+        uint256,
+        uint256 amount
+    ) public view override returns (uint256) {
+      //get token
+      IOptionToken token = IOptionToken(options);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+
+      // Get strike index, whether is long
+      AdapterStorageTypes.CharmOptionType memory optionType = adapterStorage.addressToOptionType(options);
+
+      //profit of exercising
+      try optionViews.getSellOptionCost(market, optionType.isLongToken, optionType.strikeIndex, amount) returns (uint v) {
+        return v;
+      } catch {
+        return 0;
+      }
+    }
+
+    function canExercise(
+        address options,
+        uint256,
+        uint256 amount
+    ) public view override returns (bool) {
+        //get token
+        IOptionToken token = IOptionToken(options);
+        //get market
+        IOptionMarket market = IOptionMarket(token.market());
+
+        if (block.timestamp < market.expiryTime()) {
+            return false;
+        }
+        if (exerciseProfit(options, 0, amount) > 0) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @notice Purchases the options contract.
+     * @param optionTerms is the terms of the option contract
+     * @param amount is the purchase amount in Wad units (10**18)
+     * @param maxCost is the max amount of paymentToken to be paid for the option (to avoid sandwich attacks, ...)
+     */
+    function purchase(
+        ProtocolAdapterTypes.OptionTerms calldata optionTerms,
+        uint256 amount,
+        uint256 maxCost
+    ) external payable override returns (uint256) {
+      require(
+          block.timestamp < optionTerms.expiry,
+          "Cannot purchase after expiry"
+      );
+
+      //get token address
+      address tokenAddress = lookupOToken(optionTerms);
+      if(tokenAddress == address(0)){
+        this.populateOTokenMappings();
+      }
+
+      tokenAddress = lookupOToken(optionTerms);
+
+      require(tokenAddress != address(0), "Market needs to exist!");
+
+      // Get strike index, whether is long
+      AdapterStorageTypes.CharmOptionType memory optionType = adapterStorage.addressToOptionType(tokenAddress);
+      //get token
+      IOptionToken token = IOptionToken(tokenAddress);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+      IERC20 baseToken = IERC20(optionTerms.optionType == ProtocolAdapterTypes.OptionType.Put ? optionTerms.strikeAsset : optionTerms.underlying);
+
+      uint256 amountIn;
+
+      if(baseToken.isETH()){
+        amountIn = market.buy{value: address(this).balance}(
```

Comment:
Since we are using the new `OptionRegistry` (https://github.com/charmfinance/charm-options/blob/main/contracts/periphery/OptionRegistry.sol), there is a possibility that an attacker would be able to take control of the OptionRegistry and maliciously set their own contract address.

To prevent the possibility of that contract stealing Ether, we need to revert here after a `buy` if we don't get back the correct amount of option tokens.

Something like:
```
require(token.balanceOf(address(this)) >= premium)
```

Reply:
we may get back the correct number of option tokens and still be attacked - option token address will be pulled from optionregistry as well so there may be a fake option token that is given back to us.